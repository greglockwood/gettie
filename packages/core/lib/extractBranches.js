import { shouldBeIgnored } from './utils';

export default function extractBranches(data) {
  const branches = [];
  const objects = new WeakSet();

  function handleArray(array, path) {
    for (let i = 0; i < array.length; i++) {
      if (shouldBeIgnored(array, i)) {
        continue;
      }

      path.push(i);
      branches.push([...path]);
      handleItem(array[i], path);
      path.pop();
    }

    return array;
  }

  function handleObject(obj, path) {
    for (const name in obj) {
      if (shouldBeIgnored(obj, name)) {
        continue;
      }

      path.push(name);
      branches.push([...path]);
      handleItem(obj[name], path);
      path.pop();
    }

    return obj;
  }

  function handleItem(data, path) {
    if (data) {
      if (data.toJSON) {
        data = data.toJSON();
      }

      if (objects.has(data) || shouldBeIgnored(data)) {
        return data;
      }

      if (Array.isArray(data)) {
        objects.add(data);
        data = handleArray(data, path);
      } else if (data.constructor === Object) {
        objects.add(data);
        data = handleObject(data, path);
      }
    }

    return data;
  }

  handleItem(data, []);

  return branches;
}
