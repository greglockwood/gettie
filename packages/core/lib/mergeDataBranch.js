import { shouldBeIgnored, hasOwnProperty, isObjectOrArray } from './utils';

export default function mergeDataBranch(source, dest, path) {
  let sourceCursor = source;
  let destCursor = dest;

  // don't merge branch with ignoring part
  for (const part of path) {
    if (sourceCursor == null) {
      break;
    }

    if (shouldBeIgnored(sourceCursor, part)) {
      return dest;
    }

    sourceCursor = sourceCursor[part];
  }

  sourceCursor = source;

  for (const part of path) {
    if (sourceCursor == null) {
      return dest;
    }

    if (hasOwnProperty(destCursor, part)) {
      sourceCursor = sourceCursor[part];
      destCursor = destCursor[part];
      continue;
    }

    let nextData = sourceCursor[part];

    if (isObjectOrArray(nextData)) {
      if (nextData.constructor === Object) {
        nextData = {};
      } else {
        nextData = [];
      }
    }

    destCursor[part] = nextData;
    sourceCursor = sourceCursor[part];
    destCursor = destCursor[part];
  }

  return dest;
}
