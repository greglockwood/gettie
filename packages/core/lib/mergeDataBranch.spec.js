import extractBranches from './extractBranches';
import mergeDataBranch from './mergeDataBranch';
import Gettie from './gettie';

describe('merge branches', () => {
  const obj = {
    a: 123,
    b: {
      [Symbol('foo')]: () => {},
      c: {
        d: 123,
      },
    },
    c: [1, { foo: 2 }, [3, 4]],
  };

  const objWithRecursion = {
    a: 123,
    b: {},
    c: [0, {}],
  };

  objWithRecursion.b.c = objWithRecursion;
  objWithRecursion.c[1] = objWithRecursion;

  function createTest(source, dest) {
    return () => {
      const branches = extractBranches(source);
      for (const branch of branches) {
        mergeDataBranch(source, dest, branch);
        expect(dest).toMatchSnapshot();
      }

      expect(branches).toEqual(extractBranches(dest));
    };
  }

  it('object', createTest(obj, {}));

  it('array', createTest([obj], []));

  it('recursive', createTest(objWithRecursion, {}));

  it(
    'ignoring',
    createTest(
      {
        ...obj,
        ...{
          foo: {
            bar: {
              [Gettie.ignoreSymbol]: true,
              baz: 123,
            },
          },
        },
      },
      {}
    )
  );

  it('path with ignoring', () => {
    const source = {
      foo: { bar: { [Gettie.ignoreSymbol]: true, baz: 123 } },
    };
    const dest = {};

    mergeDataBranch(source, dest, ['foo', 'bar', 'baz']);
    expect(dest).toMatchSnapshot();
  });

  it('path with nullish part', () => {
    const source = {
      foo: { bar: null },
    };
    const dest = {};

    mergeDataBranch(source, dest, ['foo', 'bar', 'baz']);
    expect(dest).toMatchSnapshot();
  });
});
