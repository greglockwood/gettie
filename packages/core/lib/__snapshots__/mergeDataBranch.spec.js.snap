// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge branches array 1`] = `
Array [
  Object {},
]
`;

exports[`merge branches array 2`] = `
Array [
  Object {
    "a": 123,
  },
]
`;

exports[`merge branches array 3`] = `
Array [
  Object {
    "a": 123,
    "b": Object {},
  },
]
`;

exports[`merge branches array 4`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {},
    },
  },
]
`;

exports[`merge branches array 5`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
  },
]
`;

exports[`merge branches array 6`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
    "c": Array [],
  },
]
`;

exports[`merge branches array 7`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
    "c": Array [
      1,
    ],
  },
]
`;

exports[`merge branches array 8`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
    "c": Array [
      1,
      Object {},
    ],
  },
]
`;

exports[`merge branches array 9`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
    "c": Array [
      1,
      Object {
        "foo": 2,
      },
    ],
  },
]
`;

exports[`merge branches array 10`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
    "c": Array [
      1,
      Object {
        "foo": 2,
      },
      Array [],
    ],
  },
]
`;

exports[`merge branches array 11`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
    "c": Array [
      1,
      Object {
        "foo": 2,
      },
      Array [
        3,
      ],
    ],
  },
]
`;

exports[`merge branches array 12`] = `
Array [
  Object {
    "a": 123,
    "b": Object {
      "c": Object {
        "d": 123,
      },
    },
    "c": Array [
      1,
      Object {
        "foo": 2,
      },
      Array [
        3,
        4,
      ],
    ],
  },
]
`;

exports[`merge branches ignoring 1`] = `
Object {
  "a": 123,
}
`;

exports[`merge branches ignoring 2`] = `
Object {
  "a": 123,
  "b": Object {},
}
`;

exports[`merge branches ignoring 3`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {},
  },
}
`;

exports[`merge branches ignoring 4`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
}
`;

exports[`merge branches ignoring 5`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [],
}
`;

exports[`merge branches ignoring 6`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
  ],
}
`;

exports[`merge branches ignoring 7`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {},
  ],
}
`;

exports[`merge branches ignoring 8`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
  ],
}
`;

exports[`merge branches ignoring 9`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
    Array [],
  ],
}
`;

exports[`merge branches ignoring 10`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
    Array [
      3,
    ],
  ],
}
`;

exports[`merge branches ignoring 11`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
    Array [
      3,
      4,
    ],
  ],
}
`;

exports[`merge branches ignoring 12`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
    Array [
      3,
      4,
    ],
  ],
  "foo": Object {},
}
`;

exports[`merge branches object 1`] = `
Object {
  "a": 123,
}
`;

exports[`merge branches object 2`] = `
Object {
  "a": 123,
  "b": Object {},
}
`;

exports[`merge branches object 3`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {},
  },
}
`;

exports[`merge branches object 4`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
}
`;

exports[`merge branches object 5`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [],
}
`;

exports[`merge branches object 6`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
  ],
}
`;

exports[`merge branches object 7`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {},
  ],
}
`;

exports[`merge branches object 8`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
  ],
}
`;

exports[`merge branches object 9`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
    Array [],
  ],
}
`;

exports[`merge branches object 10`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
    Array [
      3,
    ],
  ],
}
`;

exports[`merge branches object 11`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {
      "d": 123,
    },
  },
  "c": Array [
    1,
    Object {
      "foo": 2,
    },
    Array [
      3,
      4,
    ],
  ],
}
`;

exports[`merge branches path with ignoring 1`] = `Object {}`;

exports[`merge branches path with nullish part 1`] = `
Object {
  "foo": Object {
    "bar": null,
  },
}
`;

exports[`merge branches recursive 1`] = `
Object {
  "a": 123,
}
`;

exports[`merge branches recursive 2`] = `
Object {
  "a": 123,
  "b": Object {},
}
`;

exports[`merge branches recursive 3`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {},
  },
}
`;

exports[`merge branches recursive 4`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {},
  },
  "c": Array [],
}
`;

exports[`merge branches recursive 5`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {},
  },
  "c": Array [
    0,
  ],
}
`;

exports[`merge branches recursive 6`] = `
Object {
  "a": 123,
  "b": Object {
    "c": Object {},
  },
  "c": Array [
    0,
    Object {},
  ],
}
`;
